lib/opencensus/trace/exporters/stackdriver/converter.rb

```diff
@@ -0,0 +1,354 @@
+            proto
+          end
+
+          ##
+          # Convert a value that can be used for an attribute.
+          #
+          # @param [OpenCensus::Trace::TruncatableString, Integer, boolean]
+          #     obj Object to convert
+          # @return [Google::Devtools::Cloudtrace::V2::AttributeValue] The
+          #     generated proto
+          #
+          def convert_attribute_value obj
+            case obj
+            when OpenCensus::Trace::TruncatableString
+              TraceProtos::AttributeValue.new \
+                string_value: convert_truncatable_string(obj)
+            when Integer
+              TraceProtos::AttributeValue.new int_value: obj
+            when true, false
+              TraceProtos::AttributeValue.new bool_value: obj
+            end
+          end
+
+          ##
+          # Convert an attributes hash
+          #
+          # @param [Hash] attributes The map of attribute values to convert
+          # @param [Integer] dropped_attributes_count Number of dropped
+          # @return [Google::Devtools::Cloudtrace::V2::Attributes] The
+          #     generated proto
+          #
+          def convert_attributes attributes, dropped_attributes_count
+            attribute_map = {}
+            attributes.each do |k, v|
+              attribute_map[k] = convert_attribute_value v
+            end
+            TraceProtos::Span::Attributes.new \
+              attribute_map: attribute_map,
+              dropped_attributes_count: dropped_attributes_count
+          end
+
+          ##
+          # Convert a single stack frame as a Thread::Backtrace::Location
+          #
+          # @param [Thread::Backtrace::Location] frame The backtrace element to
+          #     convert
+          # @return [Google::Devtools::Cloudtrace::V2::StackTrace::StackFrame]
+          #     The generated proto
+          #
+          def convert_stack_frame frame
+            TraceProtos::StackTrace::StackFrame.new \
+              function_name: make_truncatable_string(frame.label),
+              file_name: make_truncatable_string(frame.path),
+              line_number: frame.lineno
+          end
+
+          ##
+          # Convert a full backtrace.
+          #
+          # @param [Array<Thread::Backtrace::Location>] backtrace The backtrace
+          #     element array to convert
+          # @param [Integer] dropped_frames_count Frames that were dropped
+          # @param [Integer] stack_trace_hash_id Hash of the data
+          # @return [Google::Devtools::Cloudtrace::V2::StackTrace] The
+          #     generated proto
+          #
+          def convert_stack_trace backtrace, dropped_frames_count,
+                                  stack_trace_hash_id
+            if @stack_trace_hash_ids[stack_trace_hash_id]
+              return TraceProtos::StackTrace.new \
+                stack_trace_hash_id: stack_trace_hash_id
+            end
+            @stack_trace_hash_ids[stack_trace_hash_id] = true
```

Comment:
If we're not saving anything other than the knowledge that we've seen this hash id before, we could use ruby's Set. Not sure if that makes it more readable or not. Also, totally not necessary.

Reply:
My rule of thumb is usually, do you need actual set operations (e.g. intersection, difference, union). If not, then meh, just use a hash.